src/main/scala/viper/silver/ast/Statement.scala

```diff
@@ -38,48 +38,22 @@ sealed trait Stmt extends Hashable with Infoed with Positioned with Transformabl
 }
 
 /** A statement that creates a new object and assigns it to a local variable. */
-case class NewStmt(lhs: LocalVar, fields: Seq[Field])(val pos: Position = NoPosition, val info: Info = NoInfo, val errT: ErrorTrafo = NoTrafos) extends Stmt {
+case class NewStmt[T <: Lhs](lhs: T, fields: Seq[Field])(val pos: Position = NoPosition, val info: Info = NoInfo, val errT: ErrorTrafo = NoTrafos) extends Stmt {
   override lazy val check : Seq[ConsistencyError] =
     (if(!Consistency.noResult(this)) Seq(ConsistencyError("Result variables are only allowed in postconditions of functions.", pos)) else Seq()) ++
     (if(!(Ref isSubtype lhs)) Seq(ConsistencyError(s"Left-hand side of New statement must be Ref type, but found ${lhs.typ}", lhs.pos)) else Seq())
-
-}
-
-/** An assignment to a field or a local variable */
-sealed trait AbstractAssign extends Stmt {
-
-  def lhs: Lhs
-
-  def rhs: Exp
-}
-
-object AbstractAssign {
-  def apply(lhs: Lhs, rhs: Exp)(pos: Position = NoPosition, info: Info = NoInfo, errT: ErrorTrafo = NoTrafos) = lhs match {
-    case l: LocalVar => LocalVarAssign(l, rhs)(pos, info, errT)
-    case l: FieldAccess => FieldAssign(l, rhs)(pos, info, errT)
-  }
-
-  def unapply(a: AbstractAssign) = Some((a.lhs, a.rhs))
-}
-
-/** An assignment to a local variable. */
-case class LocalVarAssign(lhs: LocalVar, rhs: Exp)(val pos: Position = NoPosition, val info: Info = NoInfo, val errT: ErrorTrafo = NoTrafos) extends AbstractAssign{
```

Comment:
We could keep objects LocalVarAssign and FieldAssign with apply and unapply methods to break less existing code I guess. Not sure if we should, but it would be an option.